### PR TITLE
C++ frontend Tutorial

### PR DESCRIPTION
**Pull request by goldsborough**
An end-to-end tutorial for the C++ frontend.

@soumith @ezyang maybe want to give it a quick glance?

### CODE REVIEW

File: advanced_source/cpp_frontend.rst [hunk 1]
```diff
@@ -0,0 +1,1346 @@
...
+
+  #include <torch/torch.h>
+
+  struct Net : torch::nn::Module {
+    Net(int64_t N, int64_t M) {
+      W = register_parameter("W", torch::randn({N, M}));
+      b = register_parameter("b", torch::randn(M));
+    }
+    torch::Tensor forward(torch::Tensor input) {
+      return torch::admm(b, input, W);
```
Comment (vishwakftw) [inline advanced_source/cpp_frontend.rst]: Nit: torch::addmm

File: advanced_source/cpp_frontend.rst [hunk 2]
```diff
@@ -0,0 +1,1346 @@
...
+We now have the necessary background and introduction to define the modules for
+the machine learning task we want to solve in this post. To recap: our task is
+to generate images of digits from the `MNIST dataset
+<http://yann.lecun.com/exdb/mnist/>`_. We want to use a `generative adversarial
+network (GAN)
+<https://papers.nips.cc/paper/5423-generative-adversarial-nets.pdf>`_ to solve
+this task. In particular, we'll use a `DCGAN architecture
+<https://arxiv.org/abs/1511.06434>`_ -- one of the first and simplest of its
+kind, but entirely sufficient for this task.
+
```
Comment (vishwakftw) [inline advanced_source/cpp_frontend.rst]: Pun intended?

File: advanced_source/cpp_frontend.rst [hunk 3]
```diff
@@ -0,0 +1,1346 @@
...
+What was a GAN aGAN?
```
Comment (vishwakftw) [inline advanced_source/cpp_frontend.rst]: Is it a typo or an intentional pun?

Comment (goldsborough) [reply advanced_source/cpp_frontend.rst]: a pun :)

File: advanced_source/cpp_frontend.rst (hunk 1 above)
Comment (goldsborough) [reply advanced_source/cpp_frontend.rst]: thanks

**Comment by brianjo** [comment]
Here's a direct link to the preview for easier reading. https://deploy-preview-408--pytorch-tutorials-preview.netlify.com/advanced/cpp_frontend.html

**Comment by ezyang** [comment]
It looks reasonable. Do you want editing or to ship as is?

File: advanced_source/cpp_frontend.rst [hunk 4]
```diff
@@ -0,0 +1,1346 @@
...
+
+.. code-block:: cpp
+
+  struct Net : torch::nn::Module {
+    Net(int64_t N, int64_t M)
+        : linear(register_module("linear", torch::nn::Linear(N, M))) {
+      another_bias = register_parameter("b", torch::randn(M));
+    }
+    torch::Tensor forward(torch::Tensor input) {
+      return linear(input) + another_bias;
```
[the remark this answers is not in the document]
Comment (CDitzel) [reply advanced_source/cpp_frontend.rst]: there seems to be an issue here regarding 

```suggestion
      return linear(input) + another_bias;
```
I suppose an overload of the call operator was intended here but as of now is not supported by neither the stable version nor the nightly build of libtorch. I find the overload highly desirable as it lends itself towards the python API and seems way more natural than having to call a forward function.

[the remark this answers is not in the document]
Comment (goldsborough) [reply advanced_source/cpp_frontend.rst]: I appreciate everyone trying out the examples, but there are some PRs that I am landing before this code works. The call operator is coming with https://github.com/pytorch/pytorch/pull/15831 which I'll land right now, and then the example repository, and then this PR

File: advanced_source/cpp_frontend.rst [hunk 5]
```diff
@@ -0,0 +1,1346 @@
...
+      generator->parameters(), torch::optim::AdamOptions(2e-4).beta1(0.5));
+  torch::optim::Adam discriminator_optimizer(
+      discriminator->parameters(), torch::optim::AdamOptions(2e-4).beta1(0.5));
+
+  if (kRestoreFromCheckpoint) {
+    torch::load(generator, "generator-checkpoint.pt");
+    torch::load(generator_optimizer, "generator-optimizer-checkpoint.pt");
+    torch::load(discriminator, "discriminator-checkpoint.pt");
+    torch::load(
+        discriminator_optimizer, "discriminator-optimizer-checkpoint.pt");
```
Comment (apaszke) [inline advanced_source/cpp_frontend.rst]: Just curious: how does C++ API deal with optimizers holding references to module parameters in this case? In Python one has to be careful to pickle the optimizer together with a module to keep the sharing intact, but in here you're not doing that.

Comment (goldsborough) [reply advanced_source/cpp_frontend.rst]: Tensor deserialization is in-place in the C++ serialization API, so the tensors to which the optimizers hold references should stay valid (as I see it)

**Pull request merged**